src/tests/rpp/test_flat_map.cpp

```diff
@@ -0,0 +1,139 @@
+    {
+        auto obs = rpp::source::just<TestType>(rpp::schedulers::immediate{}, 1, 2, 3);
+
+        SECTION("subscribe using flat_map with templated lambda")
+        {
+            obs | rpp::operators::flat_map([](auto v) { return rpp::source::just(v * 2); })
+                | rpp::ops::subscribe(mock.get_observer());
+            SECTION("observer obtains values from underlying observables")
+            {
+                CHECK(mock.get_received_values() == std::vector{ 2, 4, 6 });
+                CHECK(mock.get_on_completed_count() == 1);
+                CHECK(mock.get_on_error_count() == 0);
+            }
+        }
+        SECTION("subscribe using flat_map with error")
+        {
+            obs | rpp::operators::flat_map([](int) { return rpp::source::error<int>(std::make_exception_ptr(std::runtime_error{""})); })
+                | rpp::ops::subscribe(mock.get_observer());
+            SECTION("observer obtains values from underlying observables")
+            {
+                CHECK(mock.get_total_on_next_count() == 0);
+                CHECK(mock.get_on_completed_count() == 0);
+                CHECK(mock.get_on_error_count() == 1);
+            }
+        }
+        SECTION("subscribe using flat_map with empty")
+        {
+            obs | rpp::operators::flat_map([](int) { return rpp::source::empty<int>(); })
+                | rpp::ops::subscribe(mock.get_observer());
+            SECTION("observer obtains values from underlying observables")
+            {
+                CHECK(mock.get_total_on_next_count() == 0);
+                CHECK(mock.get_on_completed_count() == 1);
+                CHECK(mock.get_on_error_count() == 0);
+            }
+        }
+        SECTION("subscribe using flat_map with never")
+        {
+            obs | rpp::operators::flat_map([](int) { return rpp::source::never<int>(); })
+                | rpp::ops::subscribe(mock.get_observer());
+            SECTION("observer obtains values from underlying observables")
+            {
+                CHECK(mock.get_total_on_next_count() == 0);
+                CHECK(mock.get_on_completed_count() == 0);
+                CHECK(mock.get_on_error_count() == 0);
+            }
+        }
+        SECTION("subscribe using flat_map with empty in middle")
+        {
+            obs | rpp::operators::flat_map([](int v) {
+                    if (v == 2) 
+                        return rpp::source::empty<int>().as_dynamic();
+                    return rpp::source::just(v).as_dynamic(); 
+                  })
+                | rpp::ops::subscribe(mock.get_observer());
+            SECTION("observer obtains values from underlying observables")
+            {
+                CHECK(mock.get_received_values() == std::vector{ 1, 3 });
+                CHECK(mock.get_on_completed_count() == 1);
+                CHECK(mock.get_on_error_count() == 0);
+            }
+        }
+        SECTION("subscribe using flat_map with error in middle")
```

Comment:
Could you also add test over never in middle? to check that it is actually map + merge,  not map + concat =)

Reply:
done